### PR TITLE
SignalFx Go Changes

### PR DESCRIPTION
**Pull request by zvelobot**
The [go-signalfx](https://github.com/zvelo/go-signalfx) package is a good foundation for easily sending metrics to SignalFx (directly or via metricproxy). However, some things are not as easy as I would like.
1. It is difficult to have the reporter send a single datapoint and then forget it. It tracks all datapoints added to it and will report them every time &#x60;reporter.Report()&#x60; is called, until the datapoint is removed or that reporter is discarded. Example, say an event caused a counter to be incremented (now its value is 1). Every time &#x60;Report()&#x60; is called, that datapoint will be re-submitted to SignalFx even if the event has not occurred again. There are a few ways to approach this:
   1. First would be to add a new set of datapoints that &#x60;Reporter&#x60; tracks, something like &#x27;oneshot&#x27; (type &#x60;*DataPoints&#x60;) where the datapoints would be removed from that set after they are reported. The downside to this is that other methods like &#x60;reporter.NewCounter()&#x60; are unaware of &#x27;oneshot&#x27;. A remedy to this would be to create an enum (e.g. 0 = Retain, 1 = OneShot) that is passed to each of the &#x60;Reporter&#x60; methods so it knows where to retain them. This would break the &#x60;Reporter&#x60; api though.
   2. Another possibility is to create a new package that wraps &#x60;Reporter&#x60; and can track datapoints that should be retained and those that should be &#x27;oneshot&#x27;.
   3. One thing that should not be done is remove or reset any datapoint after &#x60;reporter.Report()&#x60; that exists in &#x60;reporter.datapoints&#x60; as that would break existing expectations about how &#x60;Reporter&#x60; works.
2. It would be nice to have &#x60;Reporter&#x60; send its data on a configurable interval (e.g. 1s, 5s, etc.) in a background goroutine. This likely belongs in a wrapper package. The goal with this is that if a client package uses an &#x60;Inc&#x60;, for example, it could simply run &#x60;Inc.Inc(1)&#x60; and not have to worry about running &#x60;reporter.Report()&#x60; later since it will send that on an interval. By doing this, a client will never block when modifying a a datapoint since it does not have to call &#x60;reporter.Report()&#x60;. This makes working with metrics much more of the &#x27;fire-and-forget&#x27; type that we are accustomed to with &#x60;StatsD&#x60;.

PIVOTAL_TRACKER_STORY_URL=https://www.pivotaltracker.com/story/show/97554198
PIVOTAL_TRACKER_STORY_TYPE=feature
PIVOTAL_TRACKER_STORY_ESTIMATE=8
PIVOTAL_TRACKER_STORY_OWNER=bobuhl
PIVOTAL_TRACKER_STORY_REQUESTER=jrubin
PIVOTAL_TRACKER_STORY_LABELS=metrics, signalfx

### CODE REVIEW

**Comment by robert-uhl** [comment]
Code review 2015-07-15 results:
- Rename DataPoint.ProtoDataPoints and set counters to zero.
- Add Reporter.Inc(metric, dimensions, delta) to handle cheap one-shots.
- Remove WithDimension.
- Extract Backgrounder; see if someone else has built such a thing
- Consult with SignalFX re. timestamp-ignorance semantics (and probably create a new story).

**Comment by zvbuhl** [comment]
Also, need to handle SignalFX's ordering requirements.

**Comment by robert-uhl** [comment]
I think that this is ready for review. I had to remove BackgroundJob entirely, replacing it with Reporter.RunInBackground. I will open a story for updating the library to ensure that timestamp ordering is preserved.

File: datapoints.go [hunk 1]
```diff
@@ -134,3 +130,22 @@ func (dps *DataPoints) Len() int {
 
 	return len(dps.datapoints)
 }
+
+// filter returns a new DataPoints structure consisting of only those
+// points where filterFunc is true
+func (dps *DataPoints) filter(filterFunc func(*DataPoint) bool) *DataPoints {
```
Comment (joshuarubin) [inline datapoints.go:136]: Is there a reason not to export `filter`?

**Comment by joshuarubin** [comment]
@zvbuhl looks good, just had the one question.

File: datapoints.go (hunk 1 above)
Comment (robert-uhl) [inline datapoints.go:136]: Only real reason is my instinctual conservatism. I think that the filter idiom can be pretty powerful for 'listy' containers. I can export it if you'd like; NBD.

**Comment by robert-uhl** [comment]
Merging per verbal discussion with Joshua.

**Pull request merged**